File: lib/data/models/check_response.dart
class CheckResponse {
  const CheckResponse({
    required this.session,
    required this.continueFlag,
    required this.books,
  });

  final String session;
  final int continueFlag;
  final Map<String, Map<String, BookSystemStatus>> books;

  factory CheckResponse.fromJson(Map<String, dynamic> json) {
    final session = json['session'] as String? ?? '';
    final continueFlag = json['continue'] as int? ?? 0;
    final booksJson = json['books'] as Map<String, dynamic>? ?? {};

    final books = <String, Map<String, BookSystemStatus>>{};
    for (final isbnEntry in booksJson.entries) {
      final isbn = isbnEntry.key;
      final systemsJson = isbnEntry.value as Map<String, dynamic>? ?? {};
      final systems = <String, BookSystemStatus>{};
      for (final systemEntry in systemsJson.entries) {
        final systemId = systemEntry.key;
        final statusJson = systemEntry.value as Map<String, dynamic>? ?? {};
        systems[systemId] = BookSystemStatus.fromJson(statusJson);
      }
      books[isbn] = systems;
    }

    return CheckResponse(
      session: session,
      continueFlag: continueFlag,
      books: books,
    );
  }
}

class BookSystemStatus {
  const BookSystemStatus({
    required this.status,
    this.reserveUrl,
    required this.libKeys,
  });

  final String status;
  final String? reserveUrl;
  final Map<String, String> libKeys;

  factory BookSystemStatus.fromJson(Map<String, dynamic> json) {
    final status = json['status'] as String? ?? '';
    final rawReserveUrl = json['reserveurl'] as String?;
    final reserveUrl = (rawReserveUrl != null && rawReserveUrl.isNotEmpty) ? rawReserveUrl : null;
    final libKeysJson = json['libkey'] as Map<String, dynamic>? ?? {};
    final libKeys = libKeysJson
        .map((key, value) => MapEntry(key, value as String? ?? ''));

    return BookSystemStatus(
      status: status,
      reserveUrl: reserveUrl,
      libKeys: libKeys,
    );
  }
}

File: src/data/models/checkResponse.ts
/**
 * Models for the Calil `/check` API response.
 * Mirrors the Dart `CheckResponse` / `BookSystemStatus`.
 *
 * The Calil API uses lowercase snake-case keys:
 * `session, continue, books, status, reserveurl, libkey`.
 */

export interface BookSystemStatus {
  readonly status: string;
  readonly reserveUrl?: string;
  readonly libKeys: Record<string, string>;
}

export interface CheckResponse {
  readonly session: string;
  readonly continueFlag: number;
  readonly books: Record<string, Record<string, BookSystemStatus>>;
}

function asRecord(value: unknown): Record<string, unknown> {
  if (value !== null && typeof value === 'object' && !Array.isArray(value)) {
    return value as Record<string, unknown>;
  }
  return {};
}

export function bookSystemStatusFromJson(
  json: Record<string, unknown>,
): BookSystemStatus {
  const status = typeof json.status === 'string' ? json.status : '';

  const rawReserveUrl = json.reserveurl;
  const reserveUrl =
    typeof rawReserveUrl === 'string' && rawReserveUrl.length > 0
      ? rawReserveUrl
      : undefined;

  const libKeysJson = asRecord(json.libkey);
  const libKeys: Record<string, string> = {};
  for (const [key, value] of Object.entries(libKeysJson)) {
    libKeys[key] = typeof value === 'string' ? value : '';
  }

  return { status, reserveUrl, libKeys };
}

export function checkResponseFromJson(
  json: Record<string, unknown>,
): CheckResponse {
  const session = typeof json.session === 'string' ? json.session : '';
  const continueFlag = typeof json.continue === 'number' ? json.continue : 0;

  const booksJson = asRecord(json.books);
  const books: Record<string, Record<string, BookSystemStatus>> = {};
  for (const [isbn, systemsValue] of Object.entries(booksJson)) {
    const systemsJson = asRecord(systemsValue);
    const systems: Record<string, BookSystemStatus> = {};
    for (const [systemId, statusValue] of Object.entries(systemsJson)) {
      systems[systemId] = bookSystemStatusFromJson(asRecord(statusValue));
    }
    books[isbn] = systems;
  }

  return { session, continueFlag, books };
}
